lint/indexOnlyLoop_checker.go

```diff
@@ -0,0 +1,81 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+)
+
+//! Detects when range loop is iterating over container which has pointer-type elements (that is, any *T or reference-like type as map or chan) using only key variable (no value) and container[key] occurs more than once inside loop body
+//
+// Suggests to use for key, v := range container form.
+//
+// @Before:
+// func closeFiles(files []*os.File) {
+//     for i := range files {
+//         if files[i] != nil {
+//            files[i].Close()
+//         }
+//     }
+// }
+//
+// @After:
+// func closeFilesSuggester(files []*os.File) {
+//     for _, f := range files {
+//         if f != nil {
+//             f.Close()
+//         }
+//     }
+// }
+
+func init() {
+	addChecker(&indexOnlyLoopChecker{}, attrExperimental)
+}
+
+type indexOnlyLoopChecker struct {
+	checkerBase
+}
+
+func (c *indexOnlyLoopChecker) VisitStmt(stmt ast.Stmt) {
+	if s, ok := stmt.(*ast.RangeStmt); ok && s.Key != nil {
+		sX := s.X.(*ast.Ident).Obj
+		var tp ast.Expr
+		switch sXField := sX.Decl.(type) {
+		case *ast.Field: // go 1.10
+			tp = sXField.Type
+		case *ast.ValueSpec: // go 1.11
+			tp = sXField.Type
+		default:
+			return
+		}
+		// sX should always be of *ast.ArrayType type
+		// cause we are in *ast.RangeStmt statement
+		sxFiledType, ok := tp.(*ast.ArrayType)
```

Comment:
You could use `*types.ArrayType`/`*types.SliceType` to get same info.

Reply:
added more detailed comment, I did't find way in current impl how `*types.SliceType` could arrive here... have some thoughts what test can be for that?

Reply:
This suggestion was for `types.Object`.
Let's merge this versions and I'll see if it can be done without `ast.Object` later.